src/Microsoft.VisualStudio.Composition/ComposedPart.cs

```diff
@@ -160,6 +160,57 @@ public IEnumerable<ComposedPartDiagnostic> Validate(IReadOnlyDictionary<Type, Ex
             }
         }
 
+        /// <summary>
+        /// Checks whether this part contains an import with <see cref="ImportCardinality.ExactlyOne"/> cardinality that has been invalidated,
+        /// which would invalidate this part as well.
+        /// </summary>
+        /// <param name="invalidPartDefnitionsSet">The set of definitions for the invalidated parts.</param>
+        /// <returns>The set of <see cref="ComposedPartDiagnostic"/> with the encountered errors.</returns>
+        public IEnumerable<ComposedPartDiagnostic> CheckForInvalidatedParts(ImmutableHashSet<ComposablePartDefinition> invalidPartDefnitionsSet)
+        {
+            foreach (var pair in this.SatisfyingExports)
+            {
+                var importDefinition = pair.Key.ImportDefinition;
+
+                bool invalidExportFound = false;
+
+                List<ExportDefinitionBinding> newSatisfyingExports = new List<ExportDefinitionBinding>();
+
+                foreach (var export in pair.Value)
+                {
+                    // If the part is invalid it should be removed from the satisfying exports.
+                    if (invalidPartDefnitionsSet.Contains(export.PartDefinition))
+                    {
+                        // Signal that we found an export to remove so that the dictionary is updated
+                        invalidExportFound = true;
+
+                        switch (importDefinition.Cardinality)
+                        {
+                            // Only report error if the cardinality is exactly one
+                            // For mutiple or optional we just remove the satisfying export
+                            case ImportCardinality.ExactlyOne:
+                                yield return new ComposedPartDiagnostic(
```

Comment:
Using `yield return` in an `IEnumerable<T>` method guarantees more allocations, as the generator method now allocates a state machine and the caller has to allocate an `IEnumerator<T>`.
Since the caller doesn't need to *progressively* enumerate these errors (i.e. it's enough to get the whole collection at once), if you change the return type to a concrete collection type (e.g. `ComposedPartDiagnostic[]`) then you can allocate the error collection in this method only if errors are actually found and otherwise return `Array.Empty<ComposedPartDiagnostic>` (which returns a shared object). The concrete collection type means the caller will be compiled to call the public `GetEnumerator()` method that returns a struct rather than having to allocate an `IEnumerator<T>` object.